### DISCUSSION TITLE
Inquiry Regarding Accessing VectorPostprocessor Object in a Material Object

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#25906] [Q&A General] [answered]
Dear MOOSE Community,

I hope this message finds you well. I am currently working on a project that involves the use of MOOSE for material modeling, and I am facing a particular challenge in accessing VectorPostprocessor Object within a material object.

My specific issue is related to the use of a VectorPostprocessor to store data in the form of `std::vector<std::vector<Real>> _need_data`. Since _need_data  is not of type [VectorPostprocessorValue](https://mooseframework.inl.gov/docs/doxygen/moose/MooseTypes_8h.html#ad71e6f71eb2930762cafe3634febf58d) , I am unable to use the `getVectorPostprocessorValue` function to directly initialize the data in Material Object. However, when I attempted to use the `getVectorPostprocessorObjectByName` function, I encountered a compilation error.

Here is the error message I received:
```bash
/home/pw-moose/projects/qinglong/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdateCopy.C: In constructor 'CrystalPlasticityKalidindiUpdateCopy::CrystalPlasticityKalidindiUpdateCopy(const InputParameters&)':
/home/pw-moose/projects/qinglong/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdateCopy.C:68:10: error: 'getVectorPostprocessorObjectByName' was not declared in this scope; did you mean 'getVectorPostprocessorValueByName'?
   68 |     _vpp(getVectorPostprocessorObjectByName("vectorpostprocessor")),
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |          getVectorPostprocessorValueByName
make: *** [/home/pw-moose/projects/moose/framework/build.mk:151: /home/pw-moose/projects/qinglong/build/unity_src/materials_crystal_plasticity_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```
If you require any additional information or code snippets to better understand the issue, please do not hesitate to ask. Your assistance in resolving this matter will be immensely valuable.

Thank you in advance for your time and support. I look forward to your response and guidance.

Best regards,

Wei

**Comment by GiudGiud** [comment]
Hello

I dont think `getVectorPostprocessorObjectByName` exists on its own.
It s a routine of the Problem so you need to do something like:
```
_fe_problem->getVectorPostprocessorObjectByName..
```

alternatively use getUserObject then dynamic cast it

Guillaume

**Reply by PengWei97** [reply]
Yes, I followed your suggestion and used getVectorPostprocessorObjectByName, but the following error is still displayed,
```bash
asticity_Unity.C:5:
/home/pw-moose/projects/qinglong/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdateCopy.C: In member function 'virtual void CrystalPlasticityKalidindiUpdateCopy::initQpStatefulProperties()':
/home/pw-moose/projects/qinglong/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdateCopy.C:106:28: error: 'const class VectorPostprocessor' has no member named 'getSlipResistance'
  106 |   auto obtain_value = _vpp.getSlipResistance(0);
      |                            ^~~~~~~~~~~~~~~~~
make: *** [/home/pw-moose/projects/moose/framework/build.mk:151: /home/pw-moose/projects/qinglong/build/unity_src/materials_crystal_plasticity_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```
The error shows that it seems that the base class VectorPostprocessor can only be called through getVectorPostprocessorObjectByName, and the derived class FeatureVolumeVectorPostprocessor of VectorPostprocessor cannot be called.

Among them, I added the following code to the material class CrystalPlasticityKalidindiUpdate, where getSlipResistance(0) is a public function in the VectorPostprocessor derived class.
```c++
  const VectorPostprocessor & vpp_object = _fe_problem.getVectorPostprocessorObjectByName("vvp_name");
  auto obtain_value = vpp_object.getSlipResistance(0);
```

wei

**Reply by GiudGiud** [reply, marked as answer]
you need to cast to get it to be the right class. 
dynamic_cast it then check that the cast succeeded and that the pointer is non-null
```
uo_ptr = dynamic_cast<FeatureVolumeVectorPostprocessor *>(&_fe_problem.getVectorPostprocessorObjectByName("vvp_name"));
if (!uo_ptr)
  mooseError("Pointer cast failed! Object is not of expected type");
```
then once you hold a pointer to a UO with the pointer type "FeatureVolumeVectorPostprocessor *" you can use it to make calls

**Reply by PengWei97** [reply]
Thank you very much for your crucial guidance, I have solved this problem, and through the following code:
```c++
auto _vpp_object_ptr = dynamic_cast<FeatureVolumeVectorPostprocessor *>(const_cast<VectorPostprocessor *>(&_fe_problem.getVectorPostprocessorObjectByName("vvp_name")));
```

**Reply by PengWei97** [reply]
Hi @GiudGiud, it's a pity that I didn't complete the last operation I wanted, even if the compilation passed. Specifically, I need to call the public member function in FeatureVolumeVectorPostprocessor::getFeatureVolume. The following is the code I added in CrystalPlasticityKalidindiUpdate::initQpStatefulProperties(), an excerpt of the input file and the error displayed when doing the dbg calculation,

CrystalPlasticityKalidindiUpdate.C
```c++
void
CrystalPlasticityKalidindiUpdateCopy::initQpStatefulProperties()
{
...
  auto _fe_problem = *parameters.get<FEProblem *>("_fe_problem");
  auto _vpp_object_ptr = dynamic_cast<FeatureVolumeVectorPostprocessor *>(const_cast<VectorPostprocessor *>(&_fe_problem.getVectorPostprocessorObjectByName("grain_volumes")));

  if (!_vpp_object_ptr)
    mooseError("Pointer cast failed! Object is not of expected type");

  std::cout << "~~~~~~~~~~~~~" << std::endl;
  unsigned int grain_id = 1;
  auto obtain_value = _vpp_object_ptr->getFeatureVolume(grain_id);  // where the error occurred
}
```

input.i
```bash
[VectorPostprocessors]
  [./grain_volumes]
    type = FeatureVolumeVectorPostprocessor
    flood_counter = grain_tracker
    execute_on = 'INITIAL TIMESTEP_BEGIN' 
  [../]
[]
``` 
error.log in dbg

```bash
Finished inside of GrainTracker

~~~~~~

Thread 1 "qinglong-opt" received signal SIGSEGV, Segmentation fault.
0x00007ffff7efb0dc in FeatureVolumeVectorPostprocessorCopy::getFeatureVolume(unsigned int) const () from /home/pw-moose/projects/qinglong/lib/libqinglong-opt.so.0
(gdb) 
```

It seems that the program is trying to access memory that is not allocated to it or access an illegal memory location, causing the operating system to terminate the execution of the program.

In general, my purpose is whether I can call the public member function `getFeatureVolume` of `FeatureVolumeVectorPostprocessor grain_volumes` in the material class `CrystalPlasticityKalidindiUpdate`.

**Reply by GiudGiud** [reply]
Segfaults are not a bad error to have when you can pin point which line is the problem with a debugger
See this page
https://mooseframework.inl.gov/application_development/debugging.html
we should be able to have a full backtrace if you cant get what is wrong from just the current entry in the backtrace `FeatureVolumeVectorPostprocessorCopy::getFeatureVolume(unsigned int)`

**Reply by PengWei97** [reply]
Great! This problem is solved, it seems to be due to the material object being created before the vectorprocessor. Finally, thank you again for your enthusiastic guidance.
